File: agenttext_api/examples/api-client-example.ts
/**
 * Example: Using the iMessage REST API with TypeScript/JavaScript
 *
 * This demonstrates how to call the REST API endpoints from an external script.
 *
 * Prerequisites:
 * 1. Start the API server: bun run api-server.ts
 * 2. Run this example: bun run examples/api-client-example.ts
 */

// API Base URL
const BASE_URL = 'http://localhost:3000'

// Helper function to make API calls
async function apiCall<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
    const response = await fetch(`${BASE_URL}${endpoint}`, {
        headers: {
            'Content-Type': 'application/json',
            ...options.headers,
        },
        ...options,
    })

    if (!response.ok) {
        const error = await response.json()
        throw new Error(`API Error: ${error.error} - ${error.message}`)
    }

    return response.json()
}

// ==================== Examples ====================

async function main() {
    console.log('🚀 iMessage REST API Client Examples\n')

    // Replace with your actual phone number or test contact
    const TEST_RECIPIENT = process.env.TEST_RECIPIENT || '+1234567890'

    try {
        // 1. Health Check
        console.log('1️⃣  Health Check')
        const health = await apiCall<{ status: string; timestamp: string }>('/health')
        console.log('   Health:', health.status)
        console.log('')

        // 2. Get API Info
        console.log('2️⃣  Get API Info')
        const info = await apiCall<any>('/info')
        console.log('   API Name:', info.name)
        console.log('   Version:', info.version)
        console.log('')

        // 3. Send a Text Message
        console.log('3️⃣  Send Text Message')
        const sendResult = await apiCall<{ sentAt: string }>('/send', {
            method: 'POST',
            body: JSON.stringify({
                to: TEST_RECIPIENT,
                content: 'Hello from the REST API! 👋'
            })
        })
        console.log('   Sent at:', sendResult.sentAt)
        console.log('')

        // 4. Get Unread Messages
        console.log('4️⃣  Get Unread Messages')
        const unread = await apiCall<{
            groups: Array<{ sender: string; messages: any[] }>
            total: number
            senderCount: number
        }>('/messages/unread')
        console.log(`   Total: ${unread.total} unread messages from ${unread.senderCount} senders`)
        for (const group of unread.groups.slice(0, 3)) {
            console.log(`   - ${group.sender}: ${group.messages.length} messages`)
        }
        console.log('')

        // 5. Query Messages
        console.log('5️⃣  Query Recent Messages')
        const messages = await apiCall<{
            messages: any[]
            total: number
            unreadCount: number
        }>('/messages?limit=5')
        console.log(`   Found ${messages.total} messages (showing ${messages.messages.length})`)
        for (const msg of messages.messages) {
            console.log(`   - ${msg.sender}: ${msg.text?.substring(0, 50) || '[attachment]'}`)
        }
        console.log('')

        // 6. List Chats
        console.log('6️⃣  List Chats')
        const chats = await apiCall<Array<{
            chatId: string
            displayName: string
            isGroup: boolean
            unreadCount: number
        }>>('/chats?limit=5')
        console.log(`   Found ${chats.length} chats:`)
        for (const chat of chats) {
            const type = chat.isGroup ? '👥 Group' : '💬 DM'
            const unread = chat.unreadCount > 0 ? ` (${chat.unreadCount} unread)` : ''
            console.log(`   ${type}: ${chat.displayName}${unread}`)
        }
        console.log('')

        // 7. Search Messages
        console.log('7️⃣  Search Messages')
        const searchTerm = 'hello'
        const searchResults = await apiCall<{
            messages: any[]
            total: number
        }>(`/messages?search=${encodeURIComponent(searchTerm)}&limit=3`)
        console.log(`   Found ${searchResults.total} messages containing "${searchTerm}"`)
        console.log('')

        // 8. List Group Chats Only
        console.log('8️⃣  List Group Chats')
        const groupChats = await apiCall<Array<{
            displayName: string
            unreadCount: number
        }>>('/chats?type=group&limit=5')
        console.log(`   Found ${groupChats.length} group chats:`)
        for (const group of groupChats) {
            console.log(`   - ${group.displayName} (${group.unreadCount} unread)`)
        }
        console.log('')

        // 9. Batch Send (uncomment to test)
        // console.log('9️⃣  Batch Send Messages')
        // const batchResults = await apiCall<Array<{
        //     to: string
        //     success: boolean
        // }>>('/send/batch', {
        //     method: 'POST',
        //     body: JSON.stringify({
        //         messages: [
        //             { to: TEST_RECIPIENT, content: 'Batch message 1' },
        //             { to: TEST_RECIPIENT, content: 'Batch message 2' }
        //         ]
        //     })
        // })
        // console.log(`   Sent ${batchResults.filter(r => r.success).length}/${batchResults.length} messages`)
        // console.log('')

        // 10. Watcher Status
        console.log('🔟 Watcher Status')
        const watcherStatus = await apiCall<{
            active: boolean
            connections: number
        }>('/watcher/status')
        console.log(`   Active: ${watcherStatus.active}`)
        console.log(`   Connections: ${watcherStatus.connections}`)
        console.log('')

        console.log('✅ All examples completed successfully!')

    } catch (error) {
        console.error('❌ Error:', error)
    }
}

// Run examples
main().catch(console.error)

File: agenttext_api/examples/api-client-example.py
#!/usr/bin/env python3
"""
Example: Using the iMessage REST API with Python

This demonstrates how to call the REST API endpoints from a Python script.

Prerequisites:
1. Install requests: pip install requests
2. Start the API server: bun run api-server.ts
3. Run this example: python examples/api-client-example.py
"""

import os
import requests
from typing import Optional

# API Base URL
BASE_URL = "http://localhost:3000"

class IMessageAPIClient:
    """Simple client for the iMessage REST API"""

    def __init__(self, base_url: str = BASE_URL):
        self.base_url = base_url

    def _request(self, method: str, endpoint: str, **kwargs):
        """Make an API request"""
        url = f"{self.base_url}{endpoint}"
        response = requests.request(method, url, **kwargs)

        if not response.ok:
            error = response.json()
            raise Exception(f"API Error: {error.get('error')} - {error.get('message')}")

        return response.json()

    def health(self):
        """Check API health"""
        return self._request("GET", "/health")

    def info(self):
        """Get API info"""
        return self._request("GET", "/info")

    def send_message(self, to: str, content):
        """Send a message"""
        return self._request("POST", "/send", json={"to": to, "content": content})

    def send_file(self, to: str, file_path: str, text: Optional[str] = None):
        """Send a file"""
        data = {"to": to, "filePath": file_path}
        if text:
            data["text"] = text
        return self._request("POST", "/send/file", json=data)

    def send_batch(self, messages: list):
        """Send multiple messages"""
        return self._request("POST", "/send/batch", json={"messages": messages})

    def get_messages(self, **filters):
        """Get messages with optional filters"""
        return self._request("GET", "/messages", params=filters)

    def get_unread_messages(self):
        """Get unread messages"""
        return self._request("GET", "/messages/unread")

    def list_chats(self, **filters):
        """List chats with optional filters"""
        return self._request("GET", "/chats", params=filters)

    def start_watcher(self):
        """Start message watcher"""
        return self._request("POST", "/watcher/start")

    def stop_watcher(self):
        """Stop message watcher"""
        return self._request("POST", "/watcher/stop")

    def watcher_status(self):
        """Get watcher status"""
        return self._request("GET", "/watcher/status")


def main():
    print("🚀 iMessage REST API Client Examples (Python)\n")

    # Initialize client
    client = IMessageAPIClient()

    # Replace with your actual phone number or test contact
    test_recipient = os.environ.get("TEST_RECIPIENT", "+1234567890")

    try:
        # 1. Health Check
        print("1️⃣  Health Check")
        health = client.health()
        print(f"   Health: {health['status']}")
        print()

        # 2. Get API Info
        print("2️⃣  Get API Info")
        info = client.info()
        print(f"   API Name: {info['name']}")
        print(f"   Version: {info['version']}")
        print()

        # 3. Send a Text Message
        print("3️⃣  Send Text Message")
        send_result = client.send_message(test_recipient, "Hello from Python! 🐍")
        print(f"   Sent at: {send_result['sentAt']}")
        print()

        # 4. Get Unread Messages
        print("4️⃣  Get Unread Messages")
        unread = client.get_unread_messages()
        print(f"   Total: {unread['total']} unread messages from {unread['senderCount']} senders")
        for group in unread['groups'][:3]:
            print(f"   - {group['sender']}: {len(group['messages'])} messages")
        print()

        # 5. Query Messages
        print("5️⃣  Query Recent Messages")
        messages = client.get_messages(limit=5)
        print(f"   Found {messages['total']} messages (showing {len(messages['messages'])})")
        for msg in messages['messages']:
            text = msg.get('text', '[attachment]')
            preview = text[:50] if text else '[attachment]'
            print(f"   - {msg['sender']}: {preview}")
        print()

        # 6. List Chats
        print("6️⃣  List Chats")
        chats = client.list_chats(limit=5)
        print(f"   Found {len(chats)} chats:")
        for chat in chats:
            chat_type = "👥 Group" if chat['isGroup'] else "💬 DM"
            unread_str = f" ({chat['unreadCount']} unread)" if chat['unreadCount'] > 0 else ""
            print(f"   {chat_type}: {chat['displayName']}{unread_str}")
        print()

        # 7. Search Messages
        print("7️⃣  Search Messages")
        search_term = "hello"
        search_results = client.get_messages(search=search_term, limit=3)
        print(f"   Found {search_results['total']} messages containing \"{search_term}\"")
        print()

        # 8. List Group Chats Only
        print("8️⃣  List Group Chats")
        group_chats = client.list_chats(type="group", limit=5)
        print(f"   Found {len(group_chats)} group chats:")
        for group in group_chats:
            print(f"   - {group['displayName']} ({group['unreadCount']} unread)")
        print()

        # 9. Batch Send (uncomment to test)
        # print("9️⃣  Batch Send Messages")
        # batch_results = client.send_batch([
        #     {"to": test_recipient, "content": "Batch message 1"},
        #     {"to": test_recipient, "content": "Batch message 2"}
        # ])
        # success_count = sum(1 for r in batch_results if r['success'])
        # print(f"   Sent {success_count}/{len(batch_results)} messages")
        # print()

        # 10. Watcher Status
        print("🔟 Watcher Status")
        watcher_status = client.watcher_status()
        print(f"   Active: {watcher_status['active']}")
        print(f"   Connections: {watcher_status['connections']}")
        print()

        print("✅ All examples completed successfully!")

    except Exception as error:
        print(f"❌ Error: {error}")


if __name__ == "__main__":
    main()
